src/System/Interactions.jl

```diff
@@ -227,45 +223,42 @@ function energy_aux(sys::System{N}, ints::Interactions, i::Int, cells, foreachbo
     if N == 0       # Dipole mode
         for cell in cells
             s = dipoles[cell, i]
-            E += energy_and_gradient_for_classical_anisotropy(s, ints.aniso.stvexp)[1]
+            E += energy_and_gradient_for_classical_anisotropy(s, ints.onsite.stvexp)[1]
         end
     else            # SU(N) mode
         for cell in cells
-            Λ = ints.aniso.matrep
+            Λ = ints.onsite.matrep
             Z = coherents[cell, i]
             E += real(dot(Z, Λ, Z))
         end
     end
 
-    # Heisenberg exchange
-    foreachbond(ints.heisen) do J, site1, site2
+    foreachbond(ints.pair) do coupling, site1, site2
         sᵢ = dipoles[site1]
         sⱼ = dipoles[site2]
-        E += J * dot(sᵢ, sⱼ)
-    end
 
-    # Quadratic exchange matrix
-    foreachbond(ints.exchange) do J, site1, site2
-        sᵢ = dipoles[site1]
-        sⱼ = dipoles[site2]
-        E += dot(sᵢ, J, sⱼ)
-    end
+        if coupling.bilin isa Float64
```

Comment:
Dynamic dispatch? Benchmark. Also if you run this in `@profview` and see red in the top row on this line, that's an easy way to check if this is a meaningful problem

Reply:
Actually scratch that, this branch can be removed entirely! `dot(v,::Float64,v)` is defined, let `dot` branch internally! :) @kbarros

Reply:
In the Julia docs there was warning that a Union{Float64, Mat3} cannot be stored on the stack, and would get heap allocated instead. That is the reason to never load coupling.bilin into a local variable, but instead put it behind the `isa` conditionals. Those statements are treated specially by the optimizer at inference time.

Reply:
After some experimentation, the changes that Sam suggested seem to work well without heap allocation. Specifically, in this PR, it is possible to do:
```
J = coupling.bilin
dot(v, J, v)
```
and the compiler can avoid heap allocation.